### PR TITLE
Remove failure paths from cn_comp_update_split()

### PR DESCRIPTION
**Pull request by nabeelmmd**
Signed-off-by: Nabeel M Mohamed <50154757+nabeelmmd@users.noreply.github.com>

### CODE REVIEW

File: lib/cn/cn_tree.c [hunk 1]
```diff
@@ -1485,98 +1485,64 @@ cn_comp_update_spill(struct cn_compaction_work *work, struct kvset **kvsets)
...
          */
         list_splice(&right->tn_kvset_list, &retired_kvsets);
         INIT_LIST_HEAD(&right->tn_kvset_list);
 
         /* Add the right half of the split kvsets to the 'right' node.
          */
-        if (nodeidv[1] != CN_TREE_INVALID_NODEID) {
-            right->tn_nodeid = nodeidv[1];
-
+        if (right) {
```
Comment (beckerg) [inline lib/cn/cn_tree.c:1527]: When would you not have a right node?

File: lib/cn/cn_tree.c [hunk 2]
```diff
@@ -1732,8 +1660,13 @@ cn_comp_commit(struct cn_compaction_work *w)
...
 
-    if (split)
-        cn_split_nodeids_get(w, nodeidv);
+    if (split) {
+        w->cw_err = cn_split_nodes_alloc(w, split_nodeidv, split_nodev);
+        if (w->cw_err) {
+            kvdb_health_error(hp, w->cw_err);
+            goto done;
+        }
+    }
```
Comment (beckerg) [inline lib/cn/cn_tree.c:1669]: Seems rather heavy handed to bring down the kvdb..  We do similar in cn_comp_compact() as well, longer term we need to try to stay online in the face transient errors and/or situations where data integrity is not in jeopardy...

File: lib/cn/node_split.c [hunk 3]
```diff
@@ -458,3 +462,79 @@ cn_split(struct cn_compaction_work *w)
...
+                nodev[i] = NULL;
+
+                return merr(ENOMEM);
+            }
+        } else {
+            /* Use the source node as the right node */
+            nodeidv[i] = w->cw_node->tn_nodeid;
+            nodev[i] = w->cw_node;
+        }
+    }
```
Comment (beckerg) [inline lib/cn/node_split.c:523]: Seems like it'd be more straightforward if you got rid of the loop...

File: lib/cn/node_split.c [hunk 4]
```diff
@@ -458,3 +462,79 @@ cn_split(struct cn_compaction_work *w)
...
+    for (int i = LEFT; i <= RIGHT && nodev[i]; i++) {
+        /* Free only the allocated left node */
+        if (i == LEFT) {
+            route_node_free(w->cw_tree->ct_route_map, nodev[i]->tn_route_node);
+            cn_node_free(nodev[i]);
+        }
+
+        nodev[i] = NULL;
+    }
+}
```
Comment (beckerg) [inline lib/cn/node_split.c:540]: loop?

File: lib/cn/route.c [hunk 5]
```diff
@@ -328,8 +319,7 @@ route_map_insert(
             link = &(*link)->rb_right;
         } else {
             log_err("dup route detected");
-            route_node_free(map, node);
-            return NULL;
+            return merr(EEXIST);
```
Comment (beckerg) [inline lib/cn/route.c:322]: Somewhere along the line this seems to have gotten perverted.  The original RB_INSERT() API returns NULL for successful insert and a ptr to the dup node if there's a duplicate.

**Review by alexttx — APPROVED** [review]

**Review by alexttx — APPROVED** [review]

**Review by beckerg — APPROVED** [review]

**Comment by beckerg** [comment]
So, looks like your s390x builds are failing too....


The following packages have unmet dependencies:
 libcurl4-openssl-dev:s390x : Depends: libcurl4:s390x (= 7.74.0-1.3+deb11u1) but it is not installable
E: Unable to correct problems, you have held broken packages.

File: lib/cn/cn_tree.c (hunk 1 above)
Comment (nabeelmmd) [reply lib/cn/cn_tree.c:1530]: This should not happen under any normal circumstances.

This check exists just to protect against a pathological case where a node split moves all the kvsets from the source node to one side either left or right.

File: lib/cn/cn_tree.c (hunk 2 above)
Comment (nabeelmmd) [reply lib/cn/cn_tree.c:1672]: Agreed, removed the health error.

File: lib/cn/node_split.c (hunk 3 above)
Comment (nabeelmmd) [reply lib/cn/node_split.c:521]: You're right. Now it looks cleaner, thanks!

File: lib/cn/node_split.c (hunk 4 above)
Comment (nabeelmmd) [reply lib/cn/node_split.c:533]: Removed.

**Comment by tristan957** [comment]
Does anyone know how to setup a chroot in order to do a cross build? For some reason libcurl4:s390x conflicts with libcurl4. I don't quite understand.

File: lib/cn/route.c (hunk 5 above)
Comment (nabeelmmd) [reply lib/cn/route.c]: I did that without knowing that it follows the semantics of the RB_INSERT() API, sorry. I've reverted it now.

**Comment by beckerg** [comment]
Been many years for me, but seems like quite a lot of work to set up a build environment?
Any idea what triggered the libcurl problem?

**Comment by tristan957** [comment]
Nabeel, before merging please squash your commits manually to generate a clean commit message.

**Pull request merged**